sweeper/sweeper.go

```
@@ -0,0 +1,191 @@
+		s.status.EndTime = time.Now().UTC()
+		s.status.Error = err
+		s.statusMu.Unlock()
+	}()
+
+	// Perform the actual sweep.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	var cacheErr error
+	go func() {
+		defer wg.Done()
+		res := s.staticSkydClient.RebuildCache()
+		<-res.ErrAvail
+		cacheErr = res.ExternErr
+	}()
+
+	// We use an independent context because we are not strictly bound to a
+	// specific API call. Also, this operation can take significant amount of
+	// time and we don't want it to fail because of a timeout.
+	ctx := context.Background()
+	dbCtx, cancel := context.WithDeadline(ctx, time.Now().UTC().Add(database.MongoDefaultTimeout))
+	defer cancel()
+
+	// Get pinned skylinks from the DB
+	dbSkylinks, err := s.staticDB.SkylinksForServer(dbCtx, s.staticServerName)
+	if err != nil {
+		err = errors.AddContext(err, "failed to fetch skylinks for server")
+		return
+	}
+	wg.Wait()
+	if cacheErr != nil {
+		err = errors.AddContext(cacheErr, "failed to rebuild skyd cache")
+		return
+	}
+
+	unknown, missing := s.staticSkydClient.DiffPinnedSkylinks(dbSkylinks)
+
+	// Remove all unknown skylink from the database.
+	var skylink skymodules.Skylink
+	for _, sl := range unknown {
+		skylink, err = database.SkylinkFromString(sl)
+		if err != nil {
+			err = errors.AddContext(err, "invalid skylink found in DB")
+			build.Critical(err)
+			continue
+		}
+		err = s.staticDB.RemoveServerFromSkylink(ctx, skylink, s.staticServerName)
+		if err != nil {
+			err = errors.AddContext(err, "failed to unpin skylink")
+			return
+		}
+	}
+	// Add all missing skylinks to the database.
+	for _, sl := range missing {
+		skylink, err = database.SkylinkFromString(sl)
+		if err != nil {
+			err = errors.AddContext(err, "invalid skylink reported by skyd")
+			build.Critical(err)
+			continue
+		}
+		err = s.staticDB.AddServerForSkylink(ctx, skylink, s.staticServerName, false)
+		if err != nil {
+			err = errors.AddContext(err, "failed to unpin skylink")
+			return
```

wouldn't we want to try all of them instead of returning on failure?  
also, update the error message here I think it's incorrect

---

The logic here was that if we fail to add a server that means we have a Mongo issue and it wouldn't make sense to continue. If we continue with the other skylinks, we'll just spam the logs with potentially dozens of thousands of lines.

The message is wrong, thanks. Fixed.